velox/connectors/hive/iceberg/IcebergSplitReader.cpp

```diff
@@ -60,15 +78,85 @@ void IcebergSplitReader::prepareSplit(
     return;
   }
 
-  createRowReader(metadataFilter);
-
   std::shared_ptr<const HiveIcebergSplit> icebergSplit =
       std::dynamic_pointer_cast<const HiveIcebergSplit>(hiveSplit_);
+  const auto& deleteFiles = icebergSplit->deleteFiles;
+
+  // Process the equality delete files to update the scan spec and remaining
+  // filters. It needs to be done after creating the Reader and before creating
+  // the RowReader.
+
+  SubfieldFilters subfieldFilters;
+  std::vector<core::TypedExprPtr> conjunctInputs;
+
+  for (const auto& deleteFile : deleteFiles) {
+    if (deleteFile.content == FileContent::kEqualityDeletes &&
+        deleteFile.recordCount > 0) {
+      // TODO: build cache of <Partition, ExprSet> to avoid repeating file
+      // parsing across partitions. Within a single partition, the splits should
+      // be with the same equality delete files and only need to be parsed once.
+      auto equalityDeleteReader = std::make_unique<EqualityDeleteFileReader>(
+          deleteFile,
+          baseFileSchema(),
+          fileHandleFactory_,
+          executor_,
+          connectorQueryCtx_,
+          hiveConfig_,
+          ioStats_,
+          runtimeStats,
+          hiveSplit_->connectorId);
+      equalityDeleteReader->readDeleteValues(subfieldFilters, conjunctInputs);
+    }
+  }
+
+  if (!subfieldFilters.empty()) {
+    originalScanSpec_ = scanSpec_->clone();
+
+    for (auto iter = subfieldFilters.begin(); iter != subfieldFilters.end();
+         iter++) {
+      auto childSpec = scanSpec_->getOrCreateChild(iter->first);
+      childSpec->addFilter(*iter->second);
+      childSpec->setSubscript(scanSpec_->children().size() - 1);
+    }
+  }
+
+  if (!conjunctInputs.empty()) {
+    core::TypedExprPtr expression =
+        std::make_shared<core::CallTypedExpr>(BOOLEAN(), conjunctInputs, "and");
+    auto deleteExprSet = expressionEvaluator_->compile(expression);
+    VELOX_CHECK_EQ(deleteExprSet->size(), 1);
+
+    if (remainingFilterExprSet_) {
+      VELOX_CHECK_EQ(remainingFilterExprSet_->size(), 1);
+
+      std::vector<std::shared_ptr<exec::Expr>> expressions;
+      expressions.push_back(remainingFilterExprSet_->expr(0));
+      expressions.push_back(deleteExprSet->expr(0));
+
+      // Back up
+      originalRemainingFilters_ = std::make_shared<ExprSet>(
+          std::vector<core::TypedExprPtr>{},
+          remainingFilterExprSet_->execCtx());
+      *originalRemainingFilters_ = *remainingFilterExprSet_;
+
+      auto newExpr = std::make_shared<exec::Expr>(
+          BOOLEAN(), std::move(expressions), "and", true, true, true);
+      remainingFilterExprSet_.reset(new exec::ExprSet(
```

Comment:
Since manipulating remaining filter is already slow path and not likely to be the final solution for multi-key deletion, let's just leave the original remaining filter unchanged and create another `ExprSet` inside `IcebergSplitReader` and evaluate it once more inside `IcebergSplitReader::next`,

Reply:
@Yuhta I updated the PR with this comment resolved. Can you please take a look to see if the change is ok? In the IcebergSplitReader::next(), I added the following at the end:
```
  // Evaluate the remaining filter deleteExprSet_ for every batch and update the
  // output vector if it reduces any rows.
  if (deleteExprSet_) {
    auto filterStartMicros = getCurrentTimeMicro();

    filterRows_.resize(output->size());
    auto rowVector = std::dynamic_pointer_cast<RowVector>(output);
    expressionEvaluator_->evaluate(
        deleteExprSet_.get(), filterRows_, *rowVector, filterResult_);
    auto numRemainingRows = exec::processFilterResults(
        filterResult_, filterRows_, filterEvalCtx_, pool_);

    if (numRemainingRows < output->size()) {
      output = exec::wrap(
          numRemainingRows, filterEvalCtx_.selectedIndices, rowVector);
    }

    totalRemainingFilterTime_.fetch_add(
        (getCurrentTimeMicro() - filterStartMicros) * 1000,
        std::memory_order_relaxed);
  }
```